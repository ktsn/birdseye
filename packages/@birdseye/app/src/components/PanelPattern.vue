<template>
  <div class="panel-pattern">
    <PanelPatternGroup
      :data="props"
      class="group"
      title="props"
      @input="$emit('input-prop', arguments[0])"
    />
    <PanelPatternGroup
      :data="data"
      class="group"
      title="data"
      @input="$emit('input-data', arguments[0])"
    />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { ComponentDataType } from '@birdseye/core'
import PanelPatternGroup from './PanelPatternGroup.vue'
import { QualifiedData } from '@/store'

function patternDataValidator(value: any[]) {
  return value.every(v => {
    return (
      typeof Array.isArray(v.type) && typeof v.name === 'string' && 'value' in v
    )
  })
}

export default Vue.extend({
  name: 'PanelPattern',

  components: {
    PanelPatternGroup
  },

  props: {
    props: {
      type: Array as () => QualifiedData[],
      required: true,
      validator: patternDataValidator
    },

    data: {
      type: Array as () => QualifiedData[],
      required: true,
      validator: patternDataValidator
    }
  }
})
</script>

<style scoped>
.group {
  border-bottom: 1px solid var(--color-border);
  padding-bottom: 20px;
  margin-bottom: 20px;
}

.group:last-child {
  border-bottom: none;
  padding-bottom: 0;
  margin-bottom: 0;
}
</style>

<birdseye lang="yml">
name: PanelPattern
patterns:
  - name: Normal
    props:
      props:
        - type:
            - string
          name: foo
          value: foo value
      data:
        - type:
            - string
            - number
          name: bar
          value: 123
        - type:
            - boolean
          name: baz
          value: true
</birdseye>
