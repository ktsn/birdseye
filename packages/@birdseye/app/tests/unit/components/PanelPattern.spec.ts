import { shallowMount } from '@vue/test-utils'
import PanelPattern from '@/components/PanelPattern.vue'

describe('PanelPattern', () => {
  const StubGroup = {
    props: ['title', 'data'],

    render(this: any, h: Function): any {
      return h('div', [
        h('p', [`title: ${this.title}`]),
        h('p', [`data: ${JSON.stringify(this.data)}`])
      ])
    }
  }

  it('renders for props and data', () => {
    const wrapper = shallowMount(PanelPattern, {
      propsData: {
        props: [
          {
            type: ['string'],
            name: 'foo',
            value: 'string value'
          }
        ],
        data: [
          {
            type: [],
            name: 'bar',
            value: true
          }
        ]
      },
      stubs: {
        PanelPatternGroup: StubGroup
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
