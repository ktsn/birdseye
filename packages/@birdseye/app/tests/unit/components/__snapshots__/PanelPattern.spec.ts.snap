// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelPattern renders for props and data 1`] = `
<div class="panel-pattern">
  <div class="group">
    <p>title: props</p>
    <p>data: [{"type":["string"],"name":"foo","value":"string value"}]</p>
  </div>
  <div class="group">
    <p>title: data</p>
    <p>data: [{"type":[],"name":"bar","value":true}]</p>
  </div>
</div>
`;
